include/openssl/hpke.h

```diff
@@ -0,0 +1,503 @@
+                  size_t *clearlen, unsigned char *clear);
+
+/**
+ * @brief generate a key pair
+ *
+ * Used for entities that will later receive HPKE values to
+ * decrypt. Only the KEM from the suite is significant here.
+ * The ``pub` output will typically be published so that
+ * others can encrypt to the private key holder using HPKE.
+ * The ``priv`` output contains the raw private value and
+ * hence is sensitive.
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param mode is the mode (currently unused)
+ * @param suite is the ciphersuite (currently unused)
+ * @param ikmlen is the length of IKM, if supplied
+ * @param ikm is IKM, if supplied
+ * @param publen is the size of the public key buffer (exact length on output)
+ * @param pub is the public value
+ * @param privlen is the size of the private key buffer (exact length on output)
+ * @param priv is the private key
+ * @return 1 for success, other for error (error returns can be non-zero)
+ */
+int OSSL_HPKE_kg(OSSL_LIB_CTX *libctx,
+                 unsigned int mode, OSSL_HPKE_SUITE suite,
+                 size_t ikmlen, unsigned char *ikm,
+                 size_t *publen, unsigned char *pub,
+                 size_t *privlen, unsigned char *priv);
+
+/**
+ * @brief generate a key pair but keep private inside API
+ *
+ * Used for entities that will later receive HPKE values to
+ * decrypt. Only the KEM from the suite is significant here.
+ * The ``pub`` output will typically be published so that
+ * others can encrypt to the private key holder using HPKE.
+ * The ``priv`` output here is in the form of an EVP_PKEY and
+ * so the raw private value need not be exposed to the
+ * application.
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param mode is the mode (currently unused)
+ * @param suite is the ciphersuite (currently unused)
+ * @param ikmlen is the length of IKM, if supplied
+ * @param ikm is IKM, if supplied
+ * @param publen is the size of the public key buffer (exact length on output)
+ * @param pub is the public value
+ * @param priv is the private key handle
+ * @return 1 for success, other for error (error returns can be non-zero)
+ */
+int OSSL_HPKE_kg_evp(OSSL_LIB_CTX *libctx,
+                     unsigned int mode, OSSL_HPKE_SUITE suite,
+                     size_t ikmlen, unsigned char *ikm,
+                     size_t *publen, unsigned char *pub,
+                     EVP_PKEY **priv);
+
+/**
+ * @brief check if a suite is supported locally
+ *
+ * @param suite is the suite to check
+ * @return 1 for success, other for error (error returns can be non-zero)
+ */
+int OSSL_HPKE_suite_check(OSSL_HPKE_SUITE suite);
+
+/**
+ * @brief: map a kem_id and a private key buffer into an EVP_PKEY
+ *
+ * Note that the buffer is expected to be some form of probably-PEM encoded
+ * private key, but could be missing the PEM header or not, and might
+ * or might not be base64 encoded. We try handle those options as best
+ * we can.
```

Comment:
Yeah, fair point. I guess what I wanted was an API that could handle import of private keys from a buffer supporting the various options for encoding (or slightly broken encoding) noted above. I don't think that exists in OpenSSL, so since I wanted it, I did it here:-) If there were an existing API that did something close to this, I'd happily use that. I guess there're a few options for what to do, and I'm open to whichever of those folk prefer:
- keep this flexibility in HPKE more or less as-is, noting it differs a bit from the approach taken by the rest of OpenSSL
- drop ``OSSL_HPKE_prbuf2evp()`` entirely and document how import exactly one form of encoded private key via PEM_read_bio_private_key using a BIO_s_mem, leading to a loss in flexibility
- propose this kind flexibility in reading private keys as a new EVP API (``EVP_PKEY_private_key_from_buf()`` or something maybe?) in a separate PR